rfcs/20250910-graph-api-dropout/README.md

```diff
@@ -0,0 +1,197 @@
+# Support Dropout Operation in Graph API
+
+## Background
+
+Dropout [[#1]][1] is a regularization technique widely adopted in deep learning
+models, including transformers and convolutional neural networks. It helps
+prevent overfitting by randomly zeroing out elements of the input tensor during
+training and scaling the outputs to maintain the expected sum. During inference,
+Dropout is typically bypassed, and the input is passed through unchanged.
+
+Mathematically, Dropout is defined as follows:
+
+**Forward (Training)**
+
+Given input tensor $X$ and dropout probability $p$:
+
+For each element $i$:
+
+> $$
+> M_i \sim \text{Bernoulli}(1-p)
+> $$
+> $$
+> Y_i = \frac{X_i \cdot M_i}{1-p}
+> $$
+
+where $M$ is the mask tensor (same shape as $X$), and $Y$ is the output tensor.
+
+**Forward (Inference)**
+
+Dropout is typically bypassed during inference:
+
+> $$
+> Y_i = X_i
+> $$
+
+**Backward**
+
+Given gradient tensor $dY$, mask tensor $M$, and dropout probability $p$:
+
+> $$
+> dX_i = \frac{dY_i \cdot M_i}{1-p}
+> $$
+
+where $dX$ is the gradient with respect to the input.
+
+## Adoption in Frameworks and Libraries
+
+**PyTorch**
+
+troch.nn.Dropout [[#2]][2] takes `p` (probability, f32) and `inplace` (bool) as attributes,
+with one input and one output. Internally, the random number generator used to
+generate the dropout mask (u8/bool) differs by device:
+
+- CPU: Uses torch.Generator backed by the C++ Mersenne Twister engine [[#3]][3],
+  which consumes a 64-bit seed.
```

Comment:
This is interesting, does it mean that PyTorch deterministic mode is not supported on CPU by default? Or that dropout has to run sequentially?

Reply:
Yes,  the random number generator used for Dropout CPU is the default PyTorch CPU RNG, which is non determinitic (evidence: [link](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/CPUGeneratorImpl.cpp#L57)), so If users run dropout (or any random op) on CPU in a single thread and set the seed, they can get deterministic results.
If they use multiple threads, the order of RNG calls can vary, so results may not be deterministic across runs, even with the same seed.